src/main/java/de/zalando/aruha/nakadi/service/subscription/zk/CuratorZkSubscriptionClient.java

```diff
@@ -0,0 +1,356 @@
+                    serializeNode(partition.getSession(), partition.getNextSession(), partition.getState()));
+        } catch (final Exception e) {
+            exceptionListener.accept(e);
+        }
+    }
+
+    @Override
+    public void incrementTopology() {
+        try {
+            log.info("Incrementing topology version");
+            final Integer curVersion = Integer.parseInt(new String(curatorFramework.getData().forPath(getSubscriptionPath("/topology")), CHARSET));
+            curatorFramework.setData().forPath(getSubscriptionPath("/topology"), String.valueOf(curVersion + 1).getBytes(CHARSET));
+        } catch (final Exception e) {
+            exceptionListener.accept(e);
+        }
+    }
+
+    @Override
+    public Partition[] listPartitions() {
+        log.info("fetching partitions information");
+        try {
+            final List<Partition> partitions = new ArrayList<>();
+            for (final String topic : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics"))) {
+                for (final String partition : curatorFramework.getChildren().forPath(getSubscriptionPath("/topics/" + topic))) {
+                    final Partition.PartitionKey key = new Partition.PartitionKey(topic, partition);
+                    partitions.add(deserializeNode(key, curatorFramework.getData().forPath(getPartitionPath(key))));
+                }
+            }
+            return partitions.toArray(new Partition[partitions.size()]);
+        } catch (final Exception e) {
+            exceptionListener.accept(e);
+            return null;
+        }
+    }
+
+    @Override
+    public Session[] listSessions() {
+        log.info("fetching sessions information");
+        final List<Session> sessions = new ArrayList<>();
+        try {
+            for (final String sessionId : curatorFramework.getChildren().forPath(getSubscriptionPath("/sessions"))) {
+                final int weight = Integer.parseInt(new String(curatorFramework.getData().forPath(getSubscriptionPath("/sessions/" + sessionId)), CHARSET));
+                sessions.add(new Session(sessionId, weight));
+            }
+            return sessions.toArray(new Session[sessions.size()]);
+        } catch (final Exception e) {
+            exceptionListener.accept(e);
+            return null;
+        }
+    }
+
+    private class ZKSessionListChanges implements ZKSubscription, Watcher {
+        private final Runnable runnable;
+        private final AtomicBoolean canceled = new AtomicBoolean(false);
+
+        private ZKSessionListChanges(final Runnable runnable) {
+            this.runnable = runnable;
+        }
+
+        final ZKSubscription init() {
+            set();
+            return this;
+        }
+
+        protected void setInternal() throws Exception {
+            curatorFramework.getChildren().usingWatcher(this).forPath(getSubscriptionPath("/clients"));
+        }
+
+        private boolean set() {
```

Comment:
Refactored and removed it. After removing exceptionListeners it was required to refresh these listeners.